### PR TITLE
Implement Azure/Azurite Blob Storage check in GetStatus

### PR DESCRIPTION
**Pull request by acltang**
https://github.com/hwsc-org/hwsc-file-transaction-svc/issues/33

Azure/Azurite automatically retries connection 3 times before returning an error. It takes a few minutes for `GetStatus` to return an `UNAVAILABLE` state to the client.

### CODE REVIEW

File: service/file_transaction_service.py [hunk 1]
```diff
@@ -16,18 +16,29 @@ def GetStatus(self, request, context):
...
-        )
+            if self.__server.get_state_locker().current_service_state == server.State.UNAVAILABLE:
+                context.set_code = grpc.StatusCode.UNAVAILABLE.value[0]
+                context.set_details = grpc.StatusCode.UNAVAILABLE.name
+
+            # Check connection to Azure Blob Storage
+            try:
+                utility.block_blob_service.get_blob_service_properties()
+            except:
+                logger.error("failed to connect to Azure Blob Storage")
```
Comment (faraonc) [inline service/file_transaction_service.py]: we need to log the stack trace.

File: service/server.py [hunk 2]
```diff
@@ -12,18 +13,27 @@ class State(Enum):
     UNAVAILABLE = 1
 
 
+class StateLocker:
+    """Synchronizes the state of file_transaction_svc"""
+    lock = rwlock.RWLockWrite()
+    current_service_state = None
+
+    def __init__(self):
+        self.current_service_state = State.AVAILABLE
```
Comment (faraonc) [inline service/server.py]: You probably do not need a constructor and just assign the state on declaration.

File: service/server.py [hunk 3]
```diff
@@ -12,18 +13,27 @@ class State(Enum):
     UNAVAILABLE = 1
 
 
+class StateLocker:
+    """Synchronizes the state of file_transaction_svc"""
+    lock = rwlock.RWLockWrite()
+    current_service_state = None
```
Comment (faraonc) [inline service/server.py]: These fields need to be private.

File: service/file_transaction_service.py (hunk 1 above)
Comment (faraonc) [inline service/file_transaction_service.py]: we need to log the stack trace

File: service/file_transaction_service.py [hunk 4]
```diff
@@ -16,18 +16,30 @@ def GetStatus(self, request, context):
...
+
+            if self.__server.get_state_locker().get_current_service_state() == server.State.UNAVAILABLE:
+                context.set_code = grpc.StatusCode.UNAVAILABLE.value[0]
+                context.set_details = grpc.StatusCode.UNAVAILABLE.name
+
+            # Check connection to Azure Blob Storage
+            try:
+                utility.block_blob_service.get_blob_service_properties()
+            except:
+                logger.exception("failed to connect to Azure Blob Storage")
```
Comment (faraonc) [inline service/file_transaction_service.py]: We need to log the stack trace.
Maybe do the following?
```
except Exception as error:
    logger.exception(error)
```

It looks like `logging` library implements that already.
https://stackoverflow.com/questions/4564559/get-exception-description-and-stack-trace-which-caused-an-exception-all-as-a-st

https://docs.python.org/3/library/logging.html#logging.Logger.exception
```
exception(msg, *args, **kwargs)
Logs a message with level ERROR on this logger. The arguments are interpreted as for debug(). Exception info is added to the logging message. This method should only be called from an exception handler.
```

Comment (acltang) [reply service/file_transaction_service.py]: The current implementation does in fact log the stack trace, although it may not be clear in the code.

**Pull request merged**